lib/ast.js

```diff
@@ -589,61 +589,64 @@ var AST_Seq = DEFNODE("Seq", "car cdr", {
         cdr: "[AST_Node] second element in sequence"
     },
     $cons: function(x, y) {
+        if (!x) {
+            throw new Error('AST_Seq.car missing');
+        }
+        if (!y) {
+            throw new Error('AST_Seq.cdr missing');
+        }
+        if (x instanceof AST_Seq) {
+            throw new Error('AST_Seq.car cannot be AST_Seq');
+        }
         var seq = new AST_Seq(x);
         seq.car = x;
         seq.cdr = y;
         return seq;
     },
     $from_array: function(array) {
         if (array.length == 0) return null;
-        if (array.length == 1) return array[0].clone();
-        var list = null;
-        for (var i = array.length; --i >= 0;) {
+        if (array.length == 1) return array[0];
+        var i = array.length - 1;
+        var list = array[i];
+        while (--i >= 0) {
             list = AST_Seq.cons(array[i], list);
         }
-        var p = list;
-        while (p) {
-            if (p.cdr && !p.cdr.cdr) {
-                p.cdr = p.cdr.car;
-                break;
-            }
-            p = p.cdr;
-        }
```

Comment:
The only time this clean-up loop manages to be useful AFAICT is against the for-loop right above, i.e. `var list = null;`